### PR TITLE
feat: Handle pagination cases where after_id does not exist

### PR DESCRIPTION
**Pull request by Emyrk**
Throw an error to the user in these cases
- Templateversions
- Workspacebuilds

User pagination does not need it as suspended users still
have rows in the database

-----

I added unit tests to confirm status code `Bad Request` for these cases. And confirmed the `users` pagination is ok and works in both the db fake + postgres.

### CODE REVIEW

File: coderd/database/databasefake/databasefake.go [hunk 1]
```diff
@@ -231,17 +231,19 @@ func (q *fakeQuerier) GetUsers(_ context.Context, params database.GetUsersParams
...
-	if len(params.Status) > 0 {
-		usersFilteredByStatus := make([]database.User, 0, len(users))
-		for i, user := range users {
-			for _, status := range params.Status {
-				if user.Status == status {
-					usersFilteredByStatus = append(usersFilteredByStatus, users[i])
-				}
+	if len(params.Status) == 0 {
+		params.Status = []database.UserStatus{database.UserStatusActive}
+	}
```
Comment (Emyrk) [inline coderd/database/databasefake/databasefake.go:236]: postgres defaults to an "active" filter.

File: coderd/workspacebuilds.go [hunk 2]
```diff
@@ -57,6 +57,24 @@ func (api *API) workspaceBuilds(rw http.ResponseWriter, r *http.Request) {
 	if !ok {
 		return
 	}
+
+	if paginationParams.AfterID != uuid.Nil {
+		// See if the record exists first. If the record does not exist, the pagination
+		// query will not work.
+		_, err := api.Database.GetWorkspaceBuildByID(r.Context(), paginationParams.AfterID)
```
Comment (dwahler) [inline coderd/workspacebuilds.go:64]: Same potential race as `templateversions.templateVersionsByTemplate`

**Review by dwahler — APPROVED** [review]

File: coderd/templateversions.go [hunk 3]
```diff
@@ -385,6 +385,23 @@ func (api *API) templateVersionsByTemplate(rw http.ResponseWriter, r *http.Reque
 		return
 	}
 
+	if paginationParams.AfterID != uuid.Nil {
+		// See if the record exists first. If the record does not exist, the pagination
+		// query will not work.
+		_, err := api.Database.GetTemplateVersionByID(r.Context(), paginationParams.AfterID)
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/templateversions.go:391]: I am dumb, @dwahler I'll put both db calls in a transaction.

File: coderd/templateversions.go [hunk 4]
```diff
@@ -385,68 +385,77 @@ func (api *API) templateVersionsByTemplate(rw http.ResponseWriter, r *http.Reque
...
+				return err
+			} else if err != nil {
+				httpapi.Write(rw, http.StatusInternalServerError, httpapi.Response{
+					Message: fmt.Sprintf("get template version at after_id: %s", err),
+				})
+				return err
+			}
+		}
+
+		versions, err := api.Database.GetTemplateVersionsByTemplateID(r.Context(), database.GetTemplateVersionsByTemplateIDParams{
```
Comment (dwahler) [inline coderd/templateversions.go:408]:
```suggestion
		versions, err := store.GetTemplateVersionsByTemplateID(r.Context(), database.GetTemplateVersionsByTemplateIDParams{
```

File: coderd/workspacebuilds.go [hunk 5]
```diff
@@ -58,39 +58,50 @@ func (api *API) workspaceBuilds(rw http.ResponseWriter, r *http.Request) {
...
+			}
+		}
+
+		req := database.GetWorkspaceBuildByWorkspaceIDParams{
+			WorkspaceID: workspace.ID,
+			AfterID:     paginationParams.AfterID,
+			OffsetOpt:   int32(paginationParams.Offset),
+			LimitOpt:    int32(paginationParams.Limit),
+		}
+		builds, err = api.Database.GetWorkspaceBuildByWorkspaceID(r.Context(), req)
```
Comment (dwahler) [inline coderd/workspacebuilds.go:88]:
```suggestion
		builds, err = store.GetWorkspaceBuildByWorkspaceID(r.Context(), req)
```

File: coderd/templateversions.go [hunk 6]
```diff
@@ -385,68 +385,77 @@ func (api *API) templateVersionsByTemplate(rw http.ResponseWriter, r *http.Reque
...
-		if err != nil && xerrors.Is(err, sql.ErrNoRows) {
-			httpapi.Write(rw, http.StatusBadRequest, httpapi.Response{
-				Message: fmt.Sprintf("record at \"after_id\" (%q) does not exists", paginationParams.AfterID.String()),
+	var err error
+	apiVersions := []codersdk.TemplateVersion{}
+	err = api.Database.InTx(func(store database.Store) error {
+		if paginationParams.AfterID != uuid.Nil {
+			// See if the record exists first. If the record does not exist, the pagination
+			// query will not work.
+			_, err := api.Database.GetTemplateVersionByID(r.Context(), paginationParams.AfterID)
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/templateversions.go:394]: :facepalm: yup. Let me bring over the linting rule I made to catch this before.

**Pull request merged**